zipkin-storage/cassandra-v1/src/main/java/zipkin2/storage/cassandra/v1/CassandraSpanConsumer.java

```diff
@@ -80,74 +97,150 @@
-    Set<Map.Entry<String, String>> autocompleteTags = new LinkedHashSet<>();
 
     List<Call<Void>> calls = new ArrayList<>();
-    for (Span v2 : rawSpans) {
-      V1Span span = converter.convert(v2);
-      // indexing occurs by timestamp, so derive one if not present.
-      long ts_micro = v2.timestampAsLong();
-      if (ts_micro == 0L) ts_micro = guessTimestamp(v2);
+    for (int i = 0; i < spanCount; i++) {
+      Span span = spans.get(i);
+      V1Span v1Span = converter.convert(span);
 
-      insertTraces.add(insertTrace.newInput(span, ByteBuffer.wrap(encoder.write(v2)), ts_micro));
+      // trace records need an insertion timestamp, so derive one if not present.
+      long ts_micro = span.timestampAsLong();
+      if (ts_micro == 0L) ts_micro = guessTimestamp(span);
 
-      if (!searchEnabled) continue;
+      insertTraces.add(
+        insertTrace.newInput(v1Span, ByteBuffer.wrap(encoder.write(span)), ts_micro));
+    }
+
+    for (InsertTrace.Input insert : insertTraces) {
+      calls.add(insertTrace.create(insert));
+    }
+
+    if (!searchEnabled) return AggregateCall.newVoidCall(calls);
+
+    // Using set or other deduplication strategies helps avoid redundant writes.
+    Set<String> insertServiceNames = new LinkedHashSet<>();
+    Set<Entry<String, String>> insertRemoteServiceNames = new LinkedHashSet<>();
+    Set<Entry<String, String>> insertSpanNames = new LinkedHashSet<>();
+    Set<Entry<String, String>> insertAutocompleteTags = new LinkedHashSet<>();
+    Indexer indexTraceIdByServiceNames = indexTraceIdByServiceName.newIndexer();
+    Indexer indexTraceIdByRemoteServiceNames = indexTraceIdByRemoteServiceName != null
+      ? indexTraceIdByRemoteServiceName.newIndexer()
+      : Indexer.NOOP;
+    Indexer indexTraceIdBySpanNames = indexTraceIdBySpanName.newIndexer();
+    Indexer indexTraceIdByAnnotations = indexTraceIdByAnnotation.newIndexer();
+
+    for (int i = 0; i < spanCount; i++) {
+      Span span = spans.get(i);
+
+      String serviceName = span.localServiceName();
+
+      // All search parameters are partitioned on service name, so we skip populating choices when
+      // this is missing.
+      if (serviceName == null) continue;
+
+      // Search in Cassandra v1 is implemented client-side, by adding rows corresponding to search
+      // parameters. For example, a search for serviceName=app&spanName=bar will look for a
+      // service_name_index row "app.foo" for the corresponding endTs and lookback. This has a few
+      // implications:
+      //
+      // * timestamps need only millis precision (as query precision is millis)
+      // * client-side indexing is pointless if there is no timestamp
+      // * populating choices for indexes never added is also pointless
+      //
+      // We floor the micros timestamp to millis and skip populating choices on zero.
+      long timestamp = 1000L * (span.timestampAsLong() / 1000L); // QueryRequest is precise to ms
```

Comment:
`timestampMillis`

Reply:
this isn't timestamp in millis it is just floored to millis

`1467676800150134L -> 1467676800150000L`